pkg/limits/frontend/ring.go

```diff
@@ -117,42 +122,69 @@ type getAssignedPartitionsResponse struct {
-	responses := make([]getAssignedPartitionsResponse, len(rs.Instances))
+	// Initialize result maps
+	highestTimestamp := make(map[int32]int64)
+	assigned := make(map[int32]string)
 
-	// Get the partitions assigned to each instance.
-	for i, instance := range rs.Instances {
-		errg.Go(func() error {
-			client, err := g.pool.GetClientFor(instance.Addr)
-			if err != nil {
-				return err
-			}
-			resp, err := client.(logproto.IngestLimitsClient).GetAssignedPartitions(ctx, &logproto.GetAssignedPartitionsRequest{})
-			if err != nil {
-				return err
+	// Track which instances need to be queried
+	toQuery := make([]ring.InstanceDesc, 0, len(rs.Instances))
+
+	// Try to get cached entries first
+	for _, instance := range rs.Instances {
+		if g.cache == nil {
+			toQuery = append(toQuery, instance)
+			continue
+		}
+
+		if cached := g.cache.Get(instance.Addr); cached != nil {
+			// Use cached partitions, but still participate in conflict resolution
+			for partition, assignedAt := range cached.Value() {
+				if t := highestTimestamp[partition]; t < assignedAt {
+					highestTimestamp[partition] = assignedAt
+					assigned[partition] = instance.Addr
+				}
 			}
-			// No need for a mutex here as responses is a "Structured variable"
-			// as described in https://go.dev/ref/spec#Variables.
-			responses[i] = getAssignedPartitionsResponse{Addr: instance.Addr, Response: resp}
-			return nil
-		})
-	}
-	if err := errg.Wait(); err != nil {
-		return nil, err
+		} else {
+			toQuery = append(toQuery, instance)
+		}
 	}
 
-	// Deduplicate the partitions. This can happen if the call to
-	// GetAssignedPartitions is interleaved with a partition rebalance, such
-	// that two or more instances claim to be the consumer of the same
-	// partition at the same time. In case of conflicts, choose the instance
-	// with the latest timestamp.
-	highestTimestamp := make(map[int32]int64)
-	assigned := make(map[int32]string)
-	for _, resp := range responses {
-		for partition, assignedAt := range resp.Response.AssignedPartitions {
-			if t := highestTimestamp[partition]; t < assignedAt {
-				highestTimestamp[partition] = assignedAt
-				assigned[partition] = resp.Addr
+	// Query uncached instances
+	if len(toQuery) > 0 {
+		errg, ctx := errgroup.WithContext(ctx)
+		responses := make([]getAssignedPartitionsResponse, len(toQuery))
+
+		for i, instance := range toQuery {
+			errg.Go(func() error {
+				client, err := g.pool.GetClientFor(instance.Addr)
+				if err != nil {
+					return err
+				}
+				resp, err := client.(logproto.IngestLimitsClient).GetAssignedPartitions(ctx, &logproto.GetAssignedPartitionsRequest{})
+				if err != nil {
+					return err
+				}
+				responses[i] = getAssignedPartitionsResponse{Addr: instance.Addr, Response: resp}
+				return nil
+			})
+		}
+		if err := errg.Wait(); err != nil {
+			return nil, err
+		}
+
+		// Process and cache new responses
+		for _, resp := range responses {
+			assignments := make(map[int32]int64)
```

Comment:
I expected we would cache the entire result `assignments` instead of the result of individual call to `GetAssignedPartitions`. Is there any advantage of doing it that way, as it seems to add quite a lot more complexity to the code? We might also want to singleflight on cache miss?

Reply:
https://pkg.go.dev/golang.org/x/sync/singleflight

Reply:
Looking back I chose caching per instance addr to have the ability to invalidate the responses per ingest limits instance. I didn't know the existence of singleflight.

Reply:
Is this to solve situations where a partition has moved between consumers? I assume you would want to invalidate the entire cache though, otherwise you'll need to find which instance used to own a partition, which instance now owns the partition, and then invalidate both of them?